File: packages/langgraph/src/token-batcher.ts
/**
 * Batches streamed token deltas into `node.token` events, at most one batch
 * per node per `intervalMs` (~30/sec by default), so a fast provider stream
 * cannot flood the wire. A single unref'd timer flushes all pending nodes;
 * `flushNode` forces a flush before a `node.finished` so ordering holds.
 *
 * (Same design as @graphmind-ai/sdk's batcher — duplicated rather than shared
 * so neither adapter depends on the other.)
 */
import type { TokenDelta } from '@graphmind-ai/client';

export type TokenBatchSink = (nodeId: string, deltas: TokenDelta[]) => void;

interface PendingBatch {
  deltas: TokenDelta[];
  /**
   * How to emit THIS node's batch. Captured per push, because the interval
   * flush fires from a bare timer callback — outside the AsyncLocalStorage run
   * context the tokens belong to. Without it, a mid-stream batch would be
   * attributed to the session's implicit run instead of its own.
   */
  sink: TokenBatchSink;
}

export class TokenBatcher {
  private readonly pending = new Map<string, PendingBatch>();
  private timer: ReturnType<typeof setTimeout> | undefined;
  private disposed = false;

  constructor(
    private readonly defaultSink: TokenBatchSink,
    private readonly intervalMs: number = 34,
  ) {}

  push(nodeId: string, delta: TokenDelta, sink?: TokenBatchSink): void {
    if (this.disposed) return;
    const batch = this.pending.get(nodeId);
    if (batch === undefined) {
      this.pending.set(nodeId, { deltas: [delta], sink: sink ?? this.defaultSink });
    } else {
      batch.deltas.push(delta);
      if (sink !== undefined) batch.sink = sink;
    }
    if (this.timer === undefined) {
      this.timer = setTimeout(() => {
        this.timer = undefined;
        this.flushAll();
      }, this.intervalMs);
      this.timer.unref?.();
    }
  }

  flushNode(nodeId: string): void {
    const batch = this.pending.get(nodeId);
    if (batch === undefined || batch.deltas.length === 0) return;
    this.pending.delete(nodeId);
    batch.sink(nodeId, batch.deltas);
  }

  /**
   * Remove and return a node's pending deltas so the CALLER emits them.
   * Used right before a `node.finished`: the caller is already inside the
   * right run context and emits synchronously, which keeps `node.token`
   * strictly ordered before the `node.finished` it belongs to.
   */
  take(nodeId: string): TokenDelta[] | undefined {
    const batch = this.pending.get(nodeId);
    if (batch === undefined || batch.deltas.length === 0) return undefined;
    this.pending.delete(nodeId);
    return batch.deltas;
  }

  flushAll(): void {
    for (const nodeId of [...this.pending.keys()]) this.flushNode(nodeId);
  }

  dispose(): void {
    if (this.disposed) return;
    if (this.timer !== undefined) clearTimeout(this.timer);
    this.timer = undefined;
    this.flushAll();
    this.disposed = true;
  }
}

File: ruby/lib/graphmind/token_batcher.rb
# frozen_string_literal: true

module Graphmind
  # Batches streamed deltas into `node.token` events.
  #
  # A streaming LLM produces tokens far faster than a canvas can paint, and one
  # WebSocket frame per token would swamp both the socket and the viewer. Every
  # delta lands in a per-node list, and a single background thread flushes them
  # on a fixed interval (~30 Hz by default) or when the node finishes.
  class TokenBatcher
    def initialize(sink, interval = 0.034)
      @sink = sink
      @interval = interval
      @mutex = Mutex.new
      @pending = {}
      @thread = nil
      @disposed = false
    end

    def push(node_id, channel, value)
      return if @disposed || value.nil? || value.empty?

      @mutex.synchronize do
        (@pending[node_id] ||= []) << { "t" => channel, "v" => value.to_s }
        ensure_thread
      end
      nil
    end

    # Flush one node's pending deltas immediately (called before node.finished
    # so a finished node never has tokens arriving after its result).
    def flush_node(node_id)
      deltas = @mutex.synchronize { @pending.delete(node_id) }
      emit(node_id, deltas)
      nil
    end

    def flush_all
      pending = @mutex.synchronize do
        snapshot = @pending
        @pending = {}
        snapshot
      end
      pending.each { |node_id, deltas| emit(node_id, deltas) }
      nil
    end

    def dispose
      @disposed = true
      thread = @mutex.synchronize do
        t = @thread
        @thread = nil
        t
      end
      thread&.kill
      flush_all
      nil
    end

    private

    # Called with the mutex held.
    def ensure_thread
      return if @thread || @disposed

      @thread = Thread.new do
        until @disposed
          sleep(@interval)
          flush_all
        end
      end
      @thread.name = "graphmind-tokens" if @thread.respond_to?(:name=)
      @thread.report_on_exception = false
    end

    def emit(node_id, deltas)
      return if deltas.nil? || deltas.empty?

      @sink.call(node_id, deltas)
    rescue StandardError
      nil
    end
  end
end
